File: Longest Alternating Subarray Problem.py
# Function to find the length of longest sublist with alternative
# Positive and Negative elements


def find_longest_sublist(A):
    """ Find the longest sublist """
    max_len = 1
    end_index = 0
    curr_len = 1
    
    for i in range(1, len(A)):
        if A[i] * A[i - 1] < 0:
            curr_len += 1
            if curr_len > max_len:
                max_len = curr_len
                end_index = i
        else:
            curr_len = 1
    
    sub_list = A[end_index - max_len + 1:end_index + 1]
    print("The Longest alternative sublist is : ", sub_list)


if __name__ == '__main__':
    A = [1, -2, 6, 4, -3, 2, -4, -3]
    find_longest_sublist(A)

File: Longest Alternating Subarray Problem.cpp
// source : https://www.techiedelight.com/longest-alternating-subarray-problem/
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std;
void print(vector<int> &v, int l, int r){
    for(int i = l; i <= r; i++){
        cout << v[i] << " ";
    }
    cout << endl;
}
void findLongestSubArray(vector<int> &v){
    int maxLen = 1;
    int endIndex = 0;
    int currLen = 1;
    for(int i = 0; i < v.size(); i++){
        if(v[i] * v[i - 1] < 0){
            currLen++;
            if(currLen > maxLen){
                maxLen = currLen;
                endIndex = i;
            }
        }
        else{
            currLen = 1;
        }
    }
    cout << "The Longest Alternative Sub-Array is : ";
    print(v, endIndex - maxLen + 1, endIndex);
}
int main(){
    int n;
    vector<int> v;
    cin >> n;
    for(int i = 0; i < n; i++){
        int a;
        cin >> a;
        v.push_back(a);
    }
    findLongestSubArray(v);
    return 0;
}

